core/drivers/stm32_bsec.c

```diff
@@ -0,0 +1,599 @@
+	uint32_t masked_val = value & BSEC_DEN_ALL_MSK;
+	unsigned int exceptions = 0;
+
+	exceptions = bsec_lock();
+
+	io_write32(bsec_base() + BSEC_DEN_OFF, value);
+
+	if ((io_read32(bsec_base() + BSEC_DEN_OFF) ^ masked_val) == 0U)
+		result = TEE_SUCCESS;
+
+	bsec_unlock(exceptions);
+
+	return result;
+}
+
+uint32_t stm32_bsec_read_debug_conf(void)
+{
+	return io_read32(bsec_base() + BSEC_DEN_OFF);
+}
+
+bool stm32_bsec_write_sr_lock(uint32_t otp_id, uint32_t value)
+{
+	uint32_t bank = otp_bank_offset(otp_id);
+	uint32_t bank_value = 0;
+	uint32_t otp_mask = BIT(otp_id & BSEC_OTP_MASK);
+	uint32_t exceptions = 0;
+
+	if (!value)
+		return false;
+
+	exceptions = bsec_lock();
+
+	bank_value = io_read32(bsec_base() + BSEC_SRLOCK_OFF + bank);
+
+	if ((bank_value & otp_mask) != value) {
+		/*
+		 * We can write 0 in all other OTP
+		 * if the lock is activated in one of other OTP.
+		 * Write 0 has no effect.
+		 */
+		io_write32(bsec_base() + BSEC_SRLOCK_OFF + bank,
+			   bank_value | otp_mask);
+	}
+
+	bsec_unlock(exceptions);
+
+	return true;
+}
+
+bool stm32_bsec_read_sr_lock(uint32_t otp_id)
+{
+	uint32_t bank = otp_bank_offset(otp_id);
+	uint32_t otp_mask = BIT(otp_id & BSEC_OTP_MASK);
+	uint32_t bank_value = io_read32(bsec_base() + BSEC_SRLOCK_OFF + bank);
+
+	return (bank_value & otp_mask) != 0U;
+}
+
+bool stm32_bsec_write_sw_lock(uint32_t otp_id, uint32_t value)
+{
+	uint32_t bank = otp_bank_offset(otp_id);
+	uint32_t otp_mask = BIT(otp_id & BSEC_OTP_MASK);
+	uint32_t bank_value = 0;
+	unsigned int exceptions = 0;
+
+	if (!value)
+		return false;
+
+	exceptions = bsec_lock();
+
+	bank_value = io_read32(bsec_base() + BSEC_SWLOCK_OFF + bank);
+
+	if ((bank_value & otp_mask) != value) {
+		/*
+		 * We can write 0 in all other OTP
+		 * if the lock is activated in one of other OTP.
+		 * Write 0 has no effect.
+		 */
+		io_write32(bsec_base() + BSEC_SWLOCK_OFF + bank,
+			   bank_value | otp_mask);
+	}
+
+	bsec_unlock(exceptions);
+
+	return true;
+}
+
+bool stm32_bsec_read_sw_lock(uint32_t otp_id)
+{
+	uint32_t bank = otp_bank_offset(otp_id);
+	uint32_t otp_mask = BIT(otp_id & BSEC_OTP_MASK);
+	uint32_t bank_value = io_read32(bsec_base() + BSEC_SWLOCK_OFF + bank);
+
+	return (bank_value & otp_mask) != 0U;
+}
+
+bool stm32_bsec_write_sp_lock(uint32_t otp_id, uint32_t value)
+{
+	uint32_t bank = otp_bank_offset(otp_id);
+	uint32_t bank_value = 0;
+	uint32_t otp_mask = BIT(otp_id & BSEC_OTP_MASK);
+	unsigned int exceptions = 0;
+
+	if (!value)
+		return false;
+
+	exceptions = bsec_lock();
+
+	bank_value = io_read32(bsec_base() + BSEC_SPLOCK_OFF + bank);
+
+	if ((bank_value & otp_mask) != value) {
+		/*
+		 * We can write 0 in all other OTP
+		 * if the lock is activated in one of other OTP.
+		 * Write 0 has no effect.
+		 */
+		io_write32(bsec_base() + BSEC_SPLOCK_OFF + bank,
+			   bank_value | otp_mask);
+	}
+
+	bsec_unlock(exceptions);
+
+	return true;
+}
+
+bool stm32_bsec_read_sp_lock(uint32_t otp_id)
+{
```

Comment:
`stm32_bsec_read_sr_lock()`, `stm32_bsec_read_sw_lock()` and `stm32_bsec_read_sp_lock()` are all the same with the exception of `BSEC_S*LOCK_OFF`